### PR TITLE
Add a test for checking that the CUDA stubs directory is not in libcaffe2_nvrts.so's RPATH or RUNPATH

### PR DESCRIPTION
**Pull request by Flamefire**
This adds only the test from #133560 as after the refactoring in #128801 CUDA stub libraries should no longer end up in the rpath.

The test verifies this.

Closes #35418

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -10783,6 +10783,25 @@ def add_neg_dim_tests():
...
+        """
+        libdir = os.path.join(os.path.dirname(torch._C.__file__), 'lib')
+        caffe2_nvrtc = os.path.join(libdir, 'libcaffe2_nvrtc.so')
+        if os.path.exists(caffe2_nvrtc):
+            output = subprocess.check_output(['objdump', '-x', caffe2_nvrtc])
+            for line in output.split(b'\n'):
+                if b'RPATH' in line or b'RUNPATH' in line:
+                    # Catch paths like:
+                    # $ORIGIN:/usr/local/cuda-10.2/lib64/stubs:/usr/local/cuda-10.2/lib64
+                    self.assertFalse(b'/stubs:' in line)
```
Comment (janeyx99) [inline test/test_torch.py:10801]: Doesn't this already check the assert in 10803?

Comment (Flamefire) [reply test/test_torch.py:10801]: No, the difference is the trailing colon, see the examples:
1. stubs is not the last entry -> has a colon afterwards
1.  stubs is the last entry, i.e. end of string -> has no colon afterwards

**Comment by janeyx99** [comment]
@malfet this change looks good to me, but I have little knowledge on RPATH/RUNPATH so grabbing your opinion.

**Review by malfet — COMMENTED** [review]
`libcaffe2_nvrtc.so` is no longer used for CUDA builds, instead we rely on so-called LazyNVRTC, see  https://github.com/pytorch/pytorch/pull/45674 so testing what libnvrtc is linked against is somewhat unimportant at this point

File: test/test_torch.py [hunk 2]
```diff
@@ -10783,6 +10783,25 @@ def add_neg_dim_tests():
...
+    @unittest.skipIf(not sys.platform.startswith('linux'), "linux-only test")
+    def test_rpath(self):
+        """
+        Make sure RPATH (or RUNPATH) in nvrtc does not contain a cuda stubs directory
+        issue gh-35418
+        """
+        libdir = os.path.join(os.path.dirname(torch._C.__file__), 'lib')
+        caffe2_nvrtc = os.path.join(libdir, 'libcaffe2_nvrtc.so')
+        if os.path.exists(caffe2_nvrtc):
+            output = subprocess.check_output(['objdump', '-x', caffe2_nvrtc])
```
Comment (malfet) [inline test/test_torch.py:10796]: This assumes that `objdump` is always installed on the system, which might not be the case...

File: test/test_torch.py [hunk 3]
```diff
@@ -10783,6 +10783,25 @@ def add_neg_dim_tests():
         assert not hasattr(TestTorch, test_name), "Duplicated test name: " + test_name
         setattr(TestTorch, test_name, make_neg_dim_test(name, tensor_arg, arg_constr, types, extra_dim))
 
+class TestRPATH(TestCase):
+    @unittest.skipIf(not sys.platform.startswith('linux'), "linux-only test")
+    def test_rpath(self):
```
Comment (malfet) [inline test/test_torch.py:10788]: Can it be renamed into something like `test_cuda_stubs_not_in_rpath`?

**Review by malfet — CHANGES_REQUESTED** [review]
Can you provide some motivation as to why this test is needed as NVRTC is no longer used for CUDA builds. Because if it is, than lots of other tests will be failing, and having one more doesn't make it clearer to developer what has happened, is it?

File: test/test_torch.py (hunk 2 above)
Comment (Flamefire) [reply test/test_torch.py:10797]: Changed the skip marker to check that

**Comment by Flamefire** [comment]
> Can you provide some motivation as to why this test is needed as NVRTC is no longer used for CUDA builds. Because if it is, than lots of other tests will be failing, and having one more doesn't make it clearer to developer what has happened, is it?

The test is needed as paths to the CUDA stub directory must not appear in the rpath or RUNPATH of any library as that would make it unusable at runtime. This should no longer happen (it did before, see the previous PR) but we better check that it stays like that. See the referenced issue #35418

What exactly do you mean by "NVRTC is no longer used"? The library that is checked here is still created and I see linking against `CUDA::nvrtc`. So it looks to me like it is still used or at least created. IIRC (the original PR is really old by now) this specific library was chosen as it is linked against many CUDA libraries. Hence no stubs in here are a good indicator that it did work.

When this test fails it should be pretty clear why, especially with the reference to the GH issue, shouldn't it? Does it need any further description?

**Pull request closed without merging**